src/core/machine-id-setup.c

```diff
@@ -207,6 +207,9 @@ int machine_id_commit(const char *root) {
         sd_id128_t id;
         int r;
 
+        /* Before doing anything, sync everything to ensure any changes by first-boot units are persisted */
+        sync();
```

Comment:
I'm not sure how?  `sync()` doesn't return errors and I don't know any equivalent syscall that does ... Maybe I'm missing something?

Reply:
ah, sync() is stupid for not returning an error. Maybe use syncfs() on /etc and /var first, and then follow-up with sync(). After all this is mostly about stuff to change in /etc and /var, and what we mostly care about. i.e.:

```c
const char *path;

/* First, explicitly sync the file systems we care about and check if it worked */
FOREACH_STRING(path, "/etc/", "/var/") {
        r = syncfs_path(path);
        if (r < 0)
                return log_errno(r, "…");
}

/* Afterwards, sync() the rest too, but we can't check the return value for these */
sync();
```

That way we'll catch the errors we really care about, and the sync the rest as well. Not perfect but as close to perfect as it can be.